src/herder/test/UpgradesTests.cpp

```diff
@@ -2484,3 +2486,135 @@ TEST_CASE("upgrade to generalized tx set", "[upgrades]")
     txSet = TxSetFrame::makeFromTransactions(txs, *app, 0, 0);
     REQUIRE(txSet->isGeneralizedTxSet());
 }
+
+TEST_CASE("upgrade to generalized tx set in network", "[upgrades][overlay]")
+{
+    if (protocolVersionIsBefore(Config::CURRENT_LEDGER_PROTOCOL_VERSION,
+                                GENERALIZED_TX_SET_PROTOCOL_VERSION))
+    {
+        return;
+    }
+    auto networkID = sha256(getTestConfig().NETWORK_PASSPHRASE);
+    auto simulation = Topologies::core(
+        4, 0.75, Simulation::OVER_LOOPBACK, networkID, [](int i) {
+            auto cfg = getTestConfig(i, Config::TESTDB_ON_DISK_SQLITE);
+            cfg.MAX_SLOTS_TO_REMEMBER = 10;
+            cfg.TESTING_UPGRADE_LEDGER_PROTOCOL_VERSION =
+                static_cast<uint32_t>(GENERALIZED_TX_SET_PROTOCOL_VERSION) - 1;
+            return cfg;
+        });
+
+    auto nodes = simulation->getNodes();
+
+    simulation->startAllNodes();
+
+    // Wait for 3 ledgers in order to get to stable closing schedule (every 5s).
+    simulation->crankUntil(
+        [&]() { return simulation->haveAllExternalized(3, 1); },
+        2 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+    auto lclCloseTime =
+        VirtualClock::from_time_t(nodes[0]
+                                      ->getLedgerManager()
+                                      .getLastClosedLedgerHeader()
+                                      .header.scpValue.closeTime);
+
+    for (auto node : nodes)
+    {
+        Upgrades::UpgradeParameters upgrades;
+        upgrades.mProtocolVersion = std::make_optional<uint32>(
+            static_cast<uint32>(GENERALIZED_TX_SET_PROTOCOL_VERSION));
+        // Upgrade to generalized tx set in 15s (4 ledgers before update is
+        // applied).
+        upgrades.mUpgradeTime = lclCloseTime + std::chrono::seconds(15);
+        node->getHerder().setUpgrades(upgrades);
+    }
+
+    auto root = TestAccount::createRoot(*nodes[0]);
+
+    int txId = 0;
+    int64_t const minBalance =
+        nodes[0]->getLedgerManager().getLastMinBalance(0);
+    auto sendTxs = [&]() {
```

Comment:
note that then you don't need to track each individual ledger like in `testLedger`. What you can do is stop one node at the beginning  (it should be on ledger 3), crank for a few ledgers past the upgrade, then restart the node and crank some more. We can check whether we're expecting a generalized ledger or not based on the ledger number when we look at closed ledgers. I think this will also make the code much simpler and easier to follow.

Reply:
As per offline discussion, changed the test to just add a new node to the network and let it catchup, so that we don't need to worry about the restart code.